core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
         }
     }
 
+    private boolean canOptimizeAddDocument(Index index) {
```

Comment:
This method is effectively  shortcut for `index.getAutoGeneratedIdTimestamp() >= 0`, with the exception of  `LOCAL_TRANSLOG_RECOVERY`. Since we always set is isRetry to true when replaying from the local translog, I'm not sure we need this? if we don't I would just convert to assertions and remove it.